### PR TITLE
refactor photo controller create for readability

### PR DESCRIPTION
**Pull request by maco**
Back when I made the horrible nested if of doom on the photos controller, we talked about eventually refactoring it for readability. Here we go. It could probably be abstracted further (item assignment and actual photo attachment lines), but this is much better already.

### CODE REVIEW

File: app/controllers/photos_controller.rb [hunk 1]
```diff
@@ -63,35 +63,20 @@ def create
...
-        else
-          flash[:alert] = "Missing id parameter"
-        end
+    if has_valid_key && has_item_id
+      item = params[:type].camelcase.constantize.find_by_id(params[:id])
+      if item && member_owns_item(item)
+        #  This syntax is weird, so just know that it means this:
+        #  @photo.harvests << item unless @photo.harvests.include?(item)
+        #  but with the correct many-to-many relationship automatically referenced
+        (@photo.send "#{params[:type]}s") << item unless (@photo.send "#{params[:type]}s").include?(item)
```
Comment (CloCkWeRX) [inline app/controllers/photos_controller.rb]: While it might seem to be anti DRY, a case statement would likely cover the purpose of valid_models + has valid_key + calling the right method.

IE:
```
collection = case params[:type]
  when 'garden'
     @photo.gardens
  when 'planting'
    @photo.plantings
  when 'harvest'
    @photo.harvests
  else
    raise "Invalid thingy"
end

collection << item unless collection.include?(item)
```

Then the whole thing could be shifted onto the photo model quite easily, rather than living in the controller - the controller only really cares that some kind of type gets added to the photo, not how sorta thing.

File: app/controllers/photos_controller.rb [hunk 2]
```diff
@@ -63,35 +63,20 @@ def create
     @photo.owner_id = current_member.id
     @photo.set_flickr_metadata
 
-    # several models can have photos. we need to know what model and the id
-    # for the entry to attach the photo to
-    valid_models = ["planting", "harvest", "garden"]
-    if params[:type]
```
Comment (pozorvlak) [inline app/controllers/photos_controller.rb]: Is it safe to drop these two checks?

File: app/controllers/photos_controller.rb (hunk 1 above)
Comment (pozorvlak) [inline app/controllers/photos_controller.rb]: I like this idea - as you say, it's a bit less DRY, but the existing metaprogramming is pretty hard to follow.

File: app/controllers/photos_controller.rb [hunk 3]
```diff
@@ -134,6 +119,22 @@ def destroy
 
   private
 
+  def valid_models
+    ["planting", "harvest", "garden"]
+  end
+
+  def has_valid_key
```
Comment (pozorvlak) [inline app/controllers/photos_controller.rb]: Oh, right, you've moved the checks into these helper methods. Sorry!

Comment (maco) [inline app/controllers/photos_controller.rb]: The other thing to ponder is concerns. I'm not sure _exactly_ how they'd
change things as I haven't tried them yet.

**Comment by pozorvlak** [comment]
I think this is an improvement as-is, but would be even better with @CloCkWeRX's magic-removal suggestion. @maco, how would you feel about that?
